runtime/domain.c

```diff
@@ -346,17 +346,57 @@ asize_t caml_norm_minor_heap_size (intnat wsize)
   return Wsize_bsize(bs);
 }
 
+/* The current minor heap layout is as follows:
+
+- A contiguous memory block of size [Minor_heap_max * Max_domains]
+  is reserved by [caml_init_domains]. The boundaries of this
+  reserved area are stored in the globals
+    [caml_minor_heaps_base]
+  and
+    [caml_minor_heaps_base_end}.
+
+- Each domain gets a reserved section of this block of size
+  [Minor_heap_max], whose boundaries are stored as
+    [domain_self->minor_heap_base]
+  and
+    [domain_self->minor_heap_base_end]
+
+- Each domain then commits a segment of size
+    [domain_state->minor_heap_wsz]
+  starting at
+    [domain_state->minor_heap_base]
+  that it actually uses.
+
+  This is done below in
+    [caml_reallocate_minor_heap]
+  which is called both at domain-initialization (by [create_domain])
+  and if a request comes to change the minor heap size.
+
+  The boundaries of this committed memory area are
+    [domain_state->young_start]
+  and
+    [domain_state->young_end].
+*/
 int caml_reallocate_minor_heap(asize_t wsize)
 {
   caml_domain_state* domain_state = Caml_state;
   CAMLassert(domain_state->young_ptr == domain_state->young_end);
 
+  CAMLassert(
+    (/* unitialized minor heap */
+      domain_state->young_start == NULL
+      && domain_state->young_end == NULL)
+    ||
+    (/* initialized minor heap */
+      domain_state->young_start == (value*)domain_self->minor_heap_area
+      && domain_state->young_end < (value*)domain_self->minor_heap_area_end));
```

Comment:
Note to self: this debug assertion is too strict (it should be `<=` not `<`): `young_end` can be equal to `minor_heap_area_end` when the minor heap size is precisely `Minor_heap_max`. I pushed this change directly in f36e5f28aae33b0204760af3307abebf6aa82219.